core/src/kura.rs

```diff
@@ -298,21 +274,28 @@ impl Kura {
 
     /// Put a block in the queue to be stored by Kura. If the queue is
     /// full, wait until there is space.
+    ///
+    /// # Errors
+    /// Return error if kura thread was closed previously
     #[allow(clippy::expect_used)]
-    pub async fn store_block_async(&self, block: VersionedCommittedBlock) {
-        self.block_sender
-            .send(block)
-            .await
-            .expect("Kura block channel is still open.");
+    pub async fn store_block_async(
+        &self,
+        block: VersionedCommittedBlock,
+    ) -> Result<(), SendError<VersionedCommittedBlock>> {
+        self.block_sender.send(block).await
     }
 
     /// Put a block in the queue to be stored by Kura. If the queue is
     /// full, block the current thread until there is space.
+    ///
+    /// # Errors
+    /// Return error if kura thread was closed previously
     #[allow(clippy::expect_used)]
-    pub fn blocking_store_block(&self, block: VersionedCommittedBlock) {
-        self.block_sender
-            .blocking_send(block)
-            .expect("Kura block channel is still open.");
+    pub fn blocking_store_block(
+        &self,
+        block: VersionedCommittedBlock,
+    ) -> Result<(), SendError<VersionedCommittedBlock>> {
```

Comment:
This would be a leak of system responsibility. I/O is separated into it's own subsystem so that the rest of the codebase can be free from I/O code. Producing blocks and consensus is sumeragi's responsibility, once the blocks exist they are fully kura's responsibility and the API should reflect that. If you want to handle the case where the kura thread has been shutdown and the channel closed, do it in this function. Emit an error with `error!();`, but sumeragi should not be aware that this has happened. Maybe you could signal to iroha to shutdown, which is what the expect was doing, although improperly.

What cases could such an error occur? Either we are shutting down iroha and the sumeragi thread will be shutdown as well so missing a block does not matter. Or the kura thread has crashed at which point our ability to reason about the problem and write code that acts properly goes to zero. In both cases the best we can do is produce an error message and exit the program.

Reply:
A little verbose but I agree.

Reply:
Good point, thanks.